### ISSUE TITLE
Process ends when all tokens are consumed

### ISSUE THREAD

**Issue by ateodorescu** [phpmentors-jp/workflower#74]
To quote from the BPMN specification: "To continue discussing how flow proceeds throughout the Process, an End Event consumes a token that had been generated from a Start Event within the same level of Process. If parallel Sequence Flows targets the End Event, then the tokens will be consumed as they arrive. All the tokens that were generated within the Process MUST be consumed by an End Event before the Process has been completed." (page 245 from "Business Process Model and Notation (BPMN), v2.0.2")

Basically we have to change the workflow so that when "end" is called we consume the tokens as they arrive. When the last one is consumed we could set the "endDate" as a property on the workflow and also have a "closed" property on the Workflow class.

At the moment "isClosed" is actually checking if all remaining tokens are "End" events.

What do you think @iteman ?

I could push a PR to solve this.

**Comment by ateodorescu** [comment]
Another quote from the BPMN 2.0 specification (page 426):
"Each Start Event that occurs creates a token on its outgoing Sequence Flows, which is followed as described by the
semantics of the other Process elements.
A Process instance is completed, if and only if the following three conditions hold:
- If the instance was created through an instantiating Parallel Gateway, then all subsequent Events (of that
Gateway) MUST have occurred.
- There is no token remaining within the Process instance.
- No Activity of the Process is still active.
For a Process instance to become completed, all tokens in that instance MUST reach an end node, i.e., a node without
outgoing Sequence Flows. A token reaching an End Event triggers the behavior associated with the Event type, e.g.,
the associated Message is sent for a Message End Event, the associated Signal is sent for a Signal End Event, and
so on. If a token reaches a Terminate End Event, the entire Process is abnormally terminated."

Maybe we should consider implementing the Terminate End Event too.

**Comment by ateodorescu** [comment]
Fixed in https://github.com/phpmentors-jp/workflower/pull/75
